### PR TITLE
Add a fixed size table of 8 * num_exutors with 16 bit entries. Use th…

### PR DESCRIPTION
**Pull request by baldersheim**
…is for mapping the first components exact.

With more components than 8x we fall abck to to using shrunk id of 8 bits as before.
This enables perfect distribution for the first 8x and then 'good enough' for the rest.
The more there are the less impact of imperfect distribution will be.

@geirst and @toregge PR

### CODE REVIEW

**Review by toregge — APPROVED** [review]

File: staging_vespalib/src/vespa/vespalib/util/sequencedtaskexecutor.cpp [hunk 1]
```diff
@@ -103,34 +105,59 @@ ExecutorStats
...
+    }
+}
+
+ISequencedTaskExecutor::ExecutorId
+SequencedTaskExecutor::getExecutorIdPerfect(uint64_t componentId) const {
+    PerfectKeyT key = componentId;
+    std::lock_guard guard(_mutex);
+    auto found = std::find(_component2IdPerfect.begin(), _component2IdPerfect.end(), key);
+    if (found == _component2IdPerfect.end()) {
+        _component2IdPerfect.push_back(key);
```
Comment (toregge) [inline staging_vespalib/src/vespa/vespalib/util/sequencedtaskexecutor.cpp:133]: Might need some protection against pushing too many elements onto `_component2IdPerfect`. If the old size is the same as `getNumExecutors() * 8 - 1` and to threads tries to get executor id for two new unique component ids, then
both might extend the vector, causing an unsafe resize (considering that it's read without lock).

Another issue is that `std::find()` in `getExecutorId()` might access uninitialized memory if `end()` returns new end before new value is written. If that uninitialized memory matches the key used for the search then we get an invalid result.

**Review by toregge — CHANGES_REQUESTED** [review]

**Comment by baldersheim** [comment]
@toregge PTAL
I hope both issues are now addressed

File: staging_vespalib/src/vespa/vespalib/util/sequencedtaskexecutor.h [hunk 2]
```diff
@@ -43,17 +43,27 @@ class SequencedTaskExecutor final : public ISequencedTaskExecutor
     const vespalib::SyncableThreadExecutor* first_executor() const;
 
 private:
+    class ValidId {
```
Comment (toregge) [inline staging_vespalib/src/vespa/vespalib/util/sequencedtaskexecutor.h:46]: consider using `std::optional<ExecutorId>` instead.

**Review by toregge — APPROVED** [review]

File: staging_vespalib/src/tests/sequencedtaskexecutor/sequencedtaskexecutor_test.cpp [hunk 3]
```diff
@@ -272,8 +277,8 @@ TEST("require that similar names gets 7/8 unique ids with 8 executors") {
     EXPECT_EQUAL(3u, four->getExecutorIdFromName("f4").getId());
     EXPECT_EQUAL(4u, four->getExecutorIdFromName("f5").getId());
     EXPECT_EQUAL(5u, four->getExecutorIdFromName("f6").getId());
-    EXPECT_EQUAL(2u, four->getExecutorIdFromName("f7").getId());
-    EXPECT_EQUAL(6u, four->getExecutorIdFromName("f8").getId());
+    EXPECT_EQUAL(6u, four->getExecutorIdFromName("f7").getId());
```
Comment (geirst) [inline staging_vespalib/src/tests/sequencedtaskexecutor/sequencedtaskexecutor_test.cpp:280]: Please update the name of the test to reflect the new reality.

File: staging_vespalib/src/vespa/vespalib/util/sequencedtaskexecutor.h (hunk 2 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/util/sequencedtaskexecutor.h]: Done

File: staging_vespalib/src/tests/sequencedtaskexecutor/sequencedtaskexecutor_test.cpp (hunk 3 above)
Comment (baldersheim) [reply staging_vespalib/src/tests/sequencedtaskexecutor/sequencedtaskexecutor_test.cpp:280]: Fixed in #20029

**Pull request merged**